File: backend/internal/shared/kernel/values/minorunits.go
// SPDX-License-Identifier: BUSL-1.1
// SPDX-FileCopyrightText: 2026 Gradion

package values

// How many minor units a currency has, and how to say an amount out loud.
//
// Money is stored as an integer count of minor units, which is the only shape
// that does not lose cents. Rendering it is where the arithmetic goes wrong,
// and it goes wrong in two opposite directions: raw, and the reader takes
// 18000000 to mean eighteen million; divided by 100, and every zero-decimal
// currency is understated a hundredfold. VND, JPY and KRW have no minor unit at
// all — ¥18,000,000 IS eighteen million yen, and `/100` turns it into 180,000.
//
// It lives HERE, beside Money, because both callers of the table are on
// different sides of a dependency edge: the offer-draft price check is in
// package compose, and the account brief is in compose/orgbrief, which compose
// imports — so orgbrief cannot import compose and the table cannot live there.
// This package is Tier-0 and importable from both.

import (
	"fmt"
	"strconv"
	"strings"
)

// currencyMinorDigits lists the ISO-4217 codes whose minor unit is not the
// usual two digits — the zero-, three- and four-digit exceptions. Most
// currencies, including EUR and USD, carry two, so the table names the
// departures and the default below carries the rest.
//
// It is a list of the departures this build knows, NOT a claim to hold every
// one. ISO also assigns codes with no minor unit at all (the precious metals,
// XDR, the test code XTS), and the standard is amended. A code missing here
// renders at two digits and is wrong for that code — which is why the tolerable
// failure is the one MinorUnitDigits documents, and why adding a code is a
// one-line change rather than a redesign.
var currencyMinorDigits = map[string]int{
	"BIF": 0, "CLP": 0, "DJF": 0, "GNF": 0, "ISK": 0, "JPY": 0, "KMF": 0,
	"KRW": 0, "PYG": 0, "RWF": 0, "UGX": 0, "UYI": 0, "VND": 0, "VUV": 0,
	"XAF": 0, "XOF": 0, "XPF": 0,
	"BHD": 3, "IQD": 3, "JOD": 3, "KWD": 3, "LYD": 3, "OMR": 3, "TND": 3,
	// Four, and both are index units a contract may legitimately price in.
	"CLF": 4, "UYW": 4,
	// ISO assigns these NO minor unit — the precious metals, the IMF's drawing
	// right, the "no currency" and test codes. Named rather than left to the
	// default, because the default is a claim about a currency that HAS a minor
	// unit, and these do not: the integer IS the amount.
	"XAU": 0, "XAG": 0, "XPT": 0, "XPD": 0, "XDR": 0, "XXX": 0, "XTS": 0,
}

// MinorUnitExceptions is a copy of the table above, for the ONE caller that
// needs to compare it rather than use it: the gate asserting that the browser's
// mirror of this table has not drifted from it
// (backend/frontendminorunits_test.go).
//
// A copy and not the map itself, because a caller holding the real one could
// change what every money figure in the product scales by, from anywhere.
func MinorUnitExceptions() map[string]int {
	out := make(map[string]int, len(currencyMinorDigits))
	for code, digits := range currencyMinorDigits {
		out[code] = digits
	}
	return out
}

// MinorUnitDigits reports how many minor-unit digits a currency code carries.
//
// A code the table does not name answers 2, and that is the right answer rather
// than a fallback: the table holds the EXCEPTIONS, so GBP, CHF, SEK, AUD, INR
// and every other ordinary currency reach this line and two digits is what they
// carry. Two is also ISO-4217's own default.
//
// Refusing an unnamed code instead would suppress the amount for nearly every
// currency on earth while admitting the two dozen exceptions — the opposite of
// the intent, since the exceptions are the ones we enumerated precisely because
// they are unusual.
//
// The residue is a code that is genuinely an exception and missing from the
// table: it renders at two digits and is wrong for that code. The repair for
// that is one line in the table above, not a guess withheld from everybody else.
func MinorUnitDigits(currency string) int {
	if digits, ok := currencyMinorDigits[strings.ToUpper(strings.TrimSpace(currency))]; ok {
		return digits
	}
	return 2
}

// MajorUnits renders an amount of minor units as the figure a person would say:
// "180000.00" for 18000000 EUR, "18000000" for the same integer in JPY.
//
// Fixed decimal places rather than a trimmed one, because the trailing zeroes
// are what say which unit this is. "180000" and "180000.00" read as the same
// number to a person and as two different claims to anything parsing them, and
// this figure's whole job is to stop a reader taking minor units for major
// ones.
//
// A negative amount keeps its sign on the front, where it belongs, rather than
// on whichever half the division happens to put it.
func MajorUnits(amountMinor int64, currency string) string {
	digits := MinorUnitDigits(currency)
	if digits == 0 {
		return strconv.FormatInt(amountMinor, 10)
	}
	scale := powerOfTen(digits)
	// The magnitude is taken as UNSIGNED, because negating an int64 does not
	// always produce a positive one: math.MinInt64 has no positive counterpart
	// and negating it yields itself, which would print a minus sign in front of
	// a negative quotient. The API admits the whole int64 range, so the one
	// value that cannot be negated is a value it can be handed.
	// The wraparound IS the mechanism, not an oversight: uint64(x) for a
	// negative x wraps to x + 2^64, and negating that in unsigned arithmetic
	// yields |x| exactly — including for math.MinInt64, whose magnitude has no
	// int64 to hold it. That is the whole reason this does not simply negate.
	sign, magnitude := "", uint64(amountMinor) // #nosec G115 -- see above: the wrap is how |MinInt64| is obtained
	if amountMinor < 0 {
		sign, magnitude = "-", -uint64(amountMinor) // #nosec G115 -- same
	}
	// scale is 10^digits for digits in {2,3,4}, so it is small and positive.
	unsigned := uint64(scale) // #nosec G115 -- a power of ten bounded by the table above
	return fmt.Sprintf("%s%d.%0*d", sign, magnitude/unsigned, digits, magnitude%unsigned)
}

// powerOfTen is 10^digits — how many minor units make one major unit.
//
// One spelling because every caller would otherwise write the same loop:
// MajorUnits to split the figure into its two halves, WholeMajorUnits to
// truncate it to the upper one, and MinorUnitScale for a caller outside this
// package that needs the multiplier itself.
//
// Named for what it computes rather than for what its callers want, because
// the alternative — minorUnitScale beside the exported MinorUnitScale — is two
// names one capital letter apart for two different signatures, which is a
// reader's problem before it is a linter's.
func powerOfTen(digits int) int64 {
	scale := int64(1)
	for range digits {
		scale *= 10
	}
	return scale
}

// MinorUnitScale is how many minor units make one major unit of the currency:
// 100 for EUR, 1000 for KWD, 1 for VND. It is the multiplier a caller converting
// a decimal figure INTO minor units needs, which MajorUnits and WholeMajorUnits
// cannot supply because they only ever divide by it.
//
// It is exported for that one shape and no other. A caller that wants to RENDER
// an amount wants MajorUnits or WholeMajorUnits instead — reaching for the scale
// to do the division by hand is how a currency's digit count comes to be right
// in this table and wrong at the surface reading it.
func MinorUnitScale(currency string) int64 {
	return powerOfTen(MinorUnitDigits(currency))
}

// WholeMajorUnits is the amount in whole major units, the fraction discarded:
// 18000000 EUR is 180000, and the same integer in VND is 18000000 because VND
// has no minor unit at all.
//
// It exists for the prose surfaces, which say an amount out loud ("€180k")
// rather than stating it, and so need the integer rather than MajorUnits'
// fixed-decimal string. It is HERE, not at those surfaces, because the division
// is the arithmetic this file owns — the three copies that each wrote `/ 100`
// themselves are how a zero-decimal currency came to be understated
// hundredfold on three surfaces at once, one of them an outbound message.
//
// Truncation, not rounding: this figure is already an approximation the caller
// abbreviates further, and rounding 999999 EUR up to "€10000" would state a
// number the record does not hold.
func WholeMajorUnits(amountMinor int64, currency string) int64 {
	return amountMinor / powerOfTen(MinorUnitDigits(currency))
}

// MinorUnits is MajorUnits' inverse: the figure a document writes ("12500.00",
// "18000000") read back as the integer count of minor units money is stored as.
// ok is false for anything that is not a plain non-negative decimal, or that
// states more fractional digits than the currency HAS — "12.345" in EUR is not
// a rounding question, it is a misread, and rounding it would invent a cent
// nobody wrote.
//
// It lives beside MajorUnits because the two share one table, and a currency
// whose digit count is corrected in that table must move both directions at
// once. A private parser somewhere else is how the two come to disagree about
// JPY.
func MinorUnits(major, currency string) (int64, bool) {
	major = strings.TrimSpace(major)
	digits := MinorUnitDigits(currency)
	// 18 integer digits is a SHAPE bound, not the overflow guard: 18 digits plus
	// a 4-decimal currency scales past int64, and what actually refuses that is
	// ParseInt failing below. The bound is here to reject a figure no document
	// states before spending a parse on it.
	if !PlainDecimal(major, 18, digits) {
		return 0, false
	}
	whole, frac, _ := strings.Cut(major, ".")
	// Pad rather than reject a short fraction: "12.5" EUR is twelve euros fifty,
	// written the way a person writes it. Padding is exact; it adds no precision
	// the document did not state.
	frac += strings.Repeat("0", digits-len(frac))
	scaled, err := strconv.ParseInt(whole+frac, 10, 64)
	if err != nil {
		return 0, false
	}
	return scaled, true
}

File: frontend/src/format/minorunits.ts
// The scale between an amount as a person types it and the integer this
// product stores.
//
// It is NOT presentation, and that is why it is its own module beside the
// formatters rather than inside them: the same table governs the write
// direction, where locale must never reach, and format.ts's own header says
// locale never flows back into storage. These functions take no locale
// deliberately — a currency's minor-unit count is a property of the currency,
// not of who is reading.
//
// It exists because the write direction had no owner at all. Eleven call sites
// spelled `Math.round(Number(amount) * 100)` and `valueMinor / 100`, which is
// right for the euro and wrong for every currency that is not two-decimal: VND,
// JPY and KRW have no minor unit, so a dong price typed as 18,000,000 was
// stored as 1,800,000,000, and a Kuwaiti dinar — three digits — was stored at a
// tenth.
//
// WHY NOT Intl. The first version of this module read the count from
// `Intl.NumberFormat(...).resolvedOptions().maximumFractionDigits`, on the
// grounds that the runtime already ships the answer and maintains it. It ships
// a DIFFERENT answer: Intl follows CLDR, which records how a currency is used,
// and the server follows ISO 4217, which records what the standard assigns.
// They disagree on ten codes, and two of them are ordinary spendable money —
// CLDR gives MGA and IRR zero digits where ISO gives two, and IQD zero where
// ISO gives three. A browser scaling by CLDR and a server scaling by ISO is the
// hundredfold disagreement this whole change exists to remove, reintroduced one
// currency over.
//
// So the table below MIRRORS shared/kernel/values/minorunits.go, exactly, and
// TestTheFrontendMinorUnitTableMatchesTheGoOne (backend/frontendminorunits_test.go)
// fails when the two drift apart in either direction. Intl is still the right
// tool for DISPLAY — grouping, symbol placement, the reader's own conventions —
// and format.ts uses it for that. It is the wrong tool for deciding an integer
// two systems must agree on.

// MINOR_UNIT_EXCEPTIONS lists the ISO 4217 codes whose minor unit is not the
// usual two digits. Most currencies, including EUR and USD, carry two, so this
// names the departures and the default below carries the rest.
//
// A code missing here renders at two digits and is wrong for that code — which
// is the tolerable failure, because refusing an unnamed code would suppress the
// amount for nearly every currency on earth while admitting the two dozen
// exceptions. Adding a code is a one-line change HERE AND IN THE GO TABLE; the
// gate refuses one without the other.
const MINOR_UNIT_EXCEPTIONS: Readonly<Record<string, number>> = {
  BIF: 0,
  CLP: 0,
  DJF: 0,
  GNF: 0,
  ISK: 0,
  JPY: 0,
  KMF: 0,
  KRW: 0,
  PYG: 0,
  RWF: 0,
  UGX: 0,
  UYI: 0,
  VND: 0,
  VUV: 0,
  XAF: 0,
  XAG: 0,
  XAU: 0,
  XDR: 0,
  XOF: 0,
  XPD: 0,
  XPF: 0,
  XPT: 0,
  XTS: 0,
  XXX: 0,
  BHD: 3,
  IQD: 3,
  JOD: 3,
  KWD: 3,
  LYD: 3,
  OMR: 3,
  TND: 3,
  CLF: 4,
  UYW: 4,
};

// minorUnitDigits reports how many minor-unit digits a currency carries.
//
// Two is ISO 4217's own default and what an unknown or malformed code answers,
// rather than throwing: a currency we cannot place is still one somebody typed
// an amount into, and refusing to scale it would store the raw digits as minor
// units.
export function minorUnitDigits(currency: string): number {
  return MINOR_UNIT_EXCEPTIONS[currency.trim().toUpperCase()] ?? 2;
}

// toMinorUnits converts a major-unit amount — what the person typed — into the
// integer the API stores, or NaN when it cannot do so exactly.
//
// It REFUSES a figure finer than its currency rather than rounding one. That is
// the rule documentextraction already stated for the same reason: "a figure
// with more decimals than the currency has is a misread, and silently dropping
// a digit is how an amount becomes wrong by an order of magnitude." A tenth of
// a cent is not a rounding question, it is a number somebody got wrong.
//
// Refusing is also what removes the double-rounding this function had. Rounding
// to a working precision and then rounding again gave answers that were wrong
// rather than merely approximate: 1.004951 EUR came out 101 cents where 100.4951
// rounds to 100, and 0.004951 came out a whole cent for less than half of one.
// With over-precise input refused there is one rounding left, and it only ever
// resolves a binary artefact.
//
// The scaling shifts a decimal STRING rather than multiplying a float, because
// `1.005 * 100` is 100.49999999999999 and a multiply would lose the cent on an
// amount stated exactly. Shifting the point in the text cannot.
//
// Halves round AWAY from zero, so a credit and a charge of the same size scale
// to the same magnitude — Math.round sends -0.5 to -0 and 0.5 to 1.
//
// NaN, rather than 0, is the refusal: a caller building a request body writes
// `amount_minor: 0` for a garbage input, and zero is a legal price. NaN
// serialises to null, which the nullable money fields take as unpriced and the
// non-nullable ones refuse — the API decides, not a silent default.
export function toMinorUnits(major: number, currency: string): number {
  if (!Number.isFinite(major)) {
    return Number.NaN;
  }
  const digits = minorUnitDigits(currency);
  // Finer than this currency can hold. Compared against the value's own
  // rounding rather than against the typed text, because the text is already
  // gone by the time a number arrives here.
  if (Number(major.toFixed(digits)) !== major) {
    return Number.NaN;
  }
  if (digits === 0) {
    return Number.isSafeInteger(major) ? major : Number.NaN;
  }
  // Enough places that the shift below never has to round a binary artefact
  // back into position; the value is already known to need no more than
  // `digits` of them.
  const text = major.toFixed(digits + 2);
  const negative = text.startsWith("-");
  const [whole, frac = ""] = (negative ? text.slice(1) : text).split(".");
  const shifted = Number(
    `${whole}${frac.slice(0, digits)}.${frac.slice(digits)}`,
  );
  const scaled = roundHalfAwayFromZero(shifted);
  const minor = negative ? -scaled : scaled;
  return Number.isSafeInteger(minor) ? minor : Number.NaN;
}

function roundHalfAwayFromZero(value: number): number {
  return value < 0 ? -Math.round(-value) : Math.round(value);
}

// toMajorUnits is the inverse, for seeding an input from a stored amount.
//
// It returns a number and not a string on purpose: the caller decides how many
// digits to show, and a currency with no minor unit must not be handed a ".00"
// that says it has one.
export function toMajorUnits(amountMinor: number, currency: string): number {
  return amountMinor / 10 ** minorUnitDigits(currency);
}
